src/libcore/iter.rs

```diff
@@ -1565,24 +1565,13 @@ pub struct Map<I, F> {
     f: F,
 }
 
-impl<I: Iterator, F, B> Map<I, F> where F: FnMut(I::Item) -> B {
-    #[inline]
-    fn do_map(&mut self, elt: Option<I::Item>) -> Option<B> {
-        match elt {
-            Some(a) => Some((self.f)(a)),
-            _ => None
-        }
-    }
-}
-
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<B, I: Iterator, F> Iterator for Map<I, F> where F: FnMut(I::Item) -> B {
     type Item = B;
 
     #[inline]
     fn next(&mut self) -> Option<B> {
-        let next = self.iter.next();
-        self.do_map(next)
+        self.iter.next().map(|a| (self.f)(a))
```

Comment:
self.f can not be moved.
if we have "impl FnMut for &FnMut", then we can use ```&self.iter.next().map(&self.f) ```

I'm curious that it is not implemented yet.
I'm also curious that fn(Arg) is not trait Fn(Arg). If it is implemented, it also helps (in other code).